src/librustc_jobserver/lib.rs

```diff
@@ -0,0 +1,292 @@
+                // forget the token so as to not drop
+                tokens.pending -= 1;
+                tokens.used += 1;
+                log::trace!("{:?} acquired real token", std::thread::current().id());
+                return;
+            }
+            tokens = self.waiter.wait(tokens).unwrap();
+        }
+    }
+
+    fn acquire_token(&self) {
+        log::trace!("{:?} acquiring token", std::thread::current().id());
+        let mut tokens = self.tokens.lock().unwrap();
+        let mut has_requested = false;
+        loop {
+            if tokens.used == 0 {
+                log::trace!("{:?} acquired implicit token", std::thread::current().id());
+                tokens.used = 1;
+                return;
+            } else if tokens.pending > 0 {
+                // forget the token so as to not drop
+                tokens.pending -= 1;
+                tokens.used += 1;
+                log::trace!("{:?} acquired real token", std::thread::current().id());
+                return;
+            }
+            if !has_requested {
+                has_requested = true;
+                self.request_token();
+            }
+            tokens = self.waiter.wait(tokens).unwrap();
+        }
+    }
+
+    fn release_token(&self) {
+        log::trace!("{:?} releasing token", std::thread::current().id());
+        let mut tokens = self.tokens.lock().unwrap();
+        tokens.used -= 1;
+        if tokens.used == 0 {
+            log::trace!("{:?} released implicit token", std::thread::current().id());
+            self.waiter.notify_one();
+        } else {
+            log::trace!("{:?} released real token", std::thread::current().id());
+            if should_notify() {
+                eprintln!(
+                    "{}",
+                    as_json(&JobserverNotification { jobserver_event: Event::Release })
+                );
+            } else {
+                GLOBAL_CLIENT.release_raw().unwrap();
+            }
+        }
+    }
+
+    pub fn request_token(&self) {
+        log::trace!("{:?} requesting token", std::thread::current().id());
+        // Just notify, don't actually acquire here.
+        notify_acquiring_token();
+        self.helper.lock().unwrap().request_token();
+    }
+}
+
+#[must_use]
+pub struct Acquired(());
+
+impl Drop for Acquired {
+    fn drop(&mut self) {
+        release_thread();
+    }
+}
+
+pub fn helper_channel() -> &'static Helper {
+    &*HELPER
+}
+
+#[derive(RustcEncodable)]
+enum Event {
+    WillAcquire,
+    Release,
+}
+
+#[derive(RustcEncodable)]
+struct JobserverNotification {
+    jobserver_event: Event,
+}
+
+// Unlike releasing tokens, there's not really a "one size fits all" approach, as we have two
+// primary ways of acquiring a token: via the helper thread, and via the acquire_thread function.
+fn notify_acquiring_token() {
+    if should_notify() {
+        eprintln!("{}", as_json(&JobserverNotification { jobserver_event: Event::WillAcquire }));
```

Comment:
After thinking on this some more I think we really do need to get some synchronization for this and fix it before merging. Standard error isn't buffered, which means that that if this doesn't write everything out in one go it could get mixed up. Additionally the trailing newline will be written separately from the message. I think all of that adds up to that we need to integrate this more into rustc's error processing rather than having this be a separate standalone crate with so much global state.